tofu: Variable validation diagnostics must mark sensitive values

apparentlymart opened the pull request:
The existing test for input variable validation with sensitive values was building on a false premise that sensitive-flagged variables always emerge from `EvalContext.GetVariableValue` as marked. In practice that's true only for input variables passed from a parent model where the value was already marked on entry, because the marking caused directly by the `sensitive = true` argument is handled later once the value has been finalized and is being reported from the main expression scope.

Since the variable validation codepath writes a not-yet-finalized value for the variable into its own HCL expression scope directly, it needs to apply the sensitive mark itself when the variable was configured as sensitive, which therefore means that the value provided to the "condition" expression is properly marked and any diagnostics generated from it will capture that marked value.

Along with fixing the existing test to have a mock `GetVariableValue` that more realistically models `BuiltinEvalContext`'s behavior, this introduces a new test specifically about the marking of the values in the diagnostic errors to help avoid potentially regressing that specific detail, since diagnostics there can potentially originate from various different parts
of the system.

This is for #2219, but will not resolve it until a similar fix is applied to the `main` branch.

## Target Release

1.8.7

## Checklist

<!-- Please check of ALL items in this list for all PRs: -->

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

<!-- If your PR contains Go code, please make sure you check off all items on this list: --> 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

cam72cam commented:
We could either merge this into main or I could cherry-pick these changes into my PR, don't have a strong preference.

apparentlymart commented:
I'm not sure what's up with the failing website build, but this PR doesn't change the website content at all so I assume it is a pre-existing problem in the `v1.8` branch. :thinking:

apparentlymart commented:
In some in-person discussion earlier we agreed that we'd let this go into `main` first and then rebase https://github.com/opentofu/opentofu/pull/2216 on top of it once merged, so I'm going to retarget this to the `main` branch and then open a separate PR to backport it to v1.8.

apparentlymart commented:
Okay... I've rebased this onto `main` and retargeted the PR to that branch.

I also removed the changelog addition from this PR, since the timing of this work means that the v1.8 changelog is the right place for this to be recorded even though we're now merging it into `main` first. I'll put the same changelog text I originally had here in the v1.8 backport PR I'll open after this is merged.
